proposals/003-skips/README.md

```diff
@@ -0,0 +1,380 @@
+variables in components, they are static. While the Spin CLI does support
+setting the same set of environment variables in all components of an app (`spin
+up --env`), it is more of a developer shortcut and not advised. It makes each
+sandbox bigger than necessary and doesn't emphasize the component model's
+strengths. See the [Spin Application Variables
+section](#spin-application-variables) for a discussion on how variables will be
+exposed to application components.
+
+#### Spin Application Variables
+
+Instead of configuring their applications with environment variables, SpinKube
+users should use [Spin application
+variables](https://developer.fermyon.com/spin/v2/variables#application-variables).
+
+These variables are declared in a Spin application manifest (`spin.toml`) and
+their values can be configured in the [`SpinApp`
+CRD](https://www.spinkube.dev/docs/spin-operator/reference/spin-app/#spinappspecvariablesindex)
+if using the environment variable provider. Application variables can also be
+dynamically updated using [Azure Key
+Vault](https://developer.fermyon.com/spin/v2/dynamic-configuration#azure-key-vault-application-variable-provider)
+or
+[Vault](https://developer.fermyon.com/spin/v2/dynamic-configuration#azure-key-vault-application-variable-provider).
+
+However, requiring that all variables be set by an administrator is likely too
+limiting for Kubernetes contexts where there are some cases in which the
+scheduler or webhooks inject environment variables into containers that
+components wish to access. For example, by default Kubernetes injects [cluster
+information environment
+variables](https://kubernetes.io/docs/concepts/containers/container-environment/#cluster-information)
+such as `KUBERNETES_SERVICE_HOST` and `KUBERNETES_SERVICE_PORT` into each
+container. Many Kubernetes libraries expect these to be available and would be
+incompatible otherwise. However, the shim should not simply inject all container
+environment variables into every component's environment. Instead, the shim will
+expose environment variables as application variable, by resetting them in the
+environment with the `SPIN_VARIABLE` prefix. It will only do this the
+application variables listed in the Spin manifest (`spin.toml`). This means that
+Spin components can get access to these variables if specifically configured in
+the Spin manifest to have access. For example, say a a component wants access to
+the `KUBERNETES_SERVICE_ADDRESS` container environment variable, it would
+configure this in the Spin.toml as follows:
+
+```toml
+[variables]
+kubernetes_service_address = { required = true }
+
+[component.example]
+source = "target/wasm32-wasi/release/example.wasm"
+allowed_outbound_hosts = []
+[component.example.variables]
+kubernetes_service_address = "{{ kubernetes_service_address }}"
+```
+
+Then, it can be accessed in the Spin app using the variables SDK:
+
+```rust
+let addr = variables::get("kubernetes_service_address")?;
```

Comment:
I know this example demonstrates how a platform can inject environment variables into applications. However, I'm curious about the implications of a Spin application being aware of its specific platform (e.g., `FOO_SERVICE_PORT` on Kubernetes as mentioned in [this doc](https://kubernetes.io/docs/concepts/containers/container-environment/#cluster-information)). Since application variables must be defined statically in the manifest file, the application must know the target services before deployment. This means application code may change based on the deployment platform.

Reply:
That is true that that only a Kubernetes specific spin app would want access to this variable and that it makes the spin app tied to a platform and service configuration (if using more particular -- `FOO` -- named services). I don't think this is a concern, though, as there will be spin apps that are specific to K8s. I do think users will potentially find it restrictive to need to know the names of the vars in advanced (rather than listing all envs and getting the ones with `SERVICE_PORT` suffixes), but that is likely an accepted limitation of Wasm -- to promote better security.

Reply:
I agree. I think the composability feature of Wasm components allows developers to author platform-neutral components for their business logic. At deplyment time, they can choose to compose with platform-specific components to form a bigger component.